test/gtest/ucp/test_ucp_sockaddr.cc

```diff
@@ -0,0 +1,115 @@
+                ucs::is_inet_addr(ifa->ifa_addr)   &&
+                ucs::is_ib_netdev(ifa->ifa_name))
+            {
+                *listen_addr = *(struct sockaddr_in*)(void*)ifa->ifa_addr;
+                listen_addr->sin_port = ucs::get_port();
+                freeifaddrs(ifaddrs);
+                return;
+            }
+        }
+        freeifaddrs(ifaddrs);
+        UCS_TEST_SKIP_R("No interface for testing");
+    }
+
+    static void scomplete_cb(void *req, ucs_status_t status)
+    {
+        ASSERT_UCS_OK(status);
+    }
+
+    static void rcomplete_cb(void *req, ucs_status_t status,
+                             ucp_tag_recv_info_t *info)
+    {
+        ASSERT_UCS_OK(status);
+    }
+
+    void tag_send(entity& from, entity& to) {
+        uint64_t send_data = ucs_generate_uuid(0);
+        void *send_req = ucp_tag_send_nb(from.ep(), &send_data, 1,
+                                         ucp_dt_make_contig(sizeof(send_data)),
+                                         1, scomplete_cb);
+        if (send_req == NULL) {
+        } else if (UCS_PTR_IS_ERR(send_req)) {
+            ASSERT_UCS_OK(UCS_PTR_STATUS(send_req));
+        } else {
+            while (!ucp_request_is_completed(send_req)) {
+                progress();
+            }
+            ucp_request_free(send_req);
+        }
+
+        uint64_t recv_data = 0;
+        void *recv_req = ucp_tag_recv_nb(to.worker(), &recv_data, 1,
+                                         ucp_dt_make_contig(sizeof(recv_data)),
+                                         1, 0, rcomplete_cb);
+        if (UCS_PTR_IS_ERR(recv_req)) {
+            ASSERT_UCS_OK(UCS_PTR_STATUS(recv_req));
+        } else {
+            while (!ucp_request_is_completed(recv_req)) {
+                progress();
+            }
+            ucp_request_free(recv_req);
+        }
+
+        EXPECT_EQ(send_data, recv_data);
+    }
+};
+
+UCS_TEST_P(test_ucp_sockaddr, listen) {
+
+    struct sockaddr_in listen_addr;
+    get_listen_addr(&listen_addr);
+
+    ucs_status_t status = receiver().listen((const struct sockaddr*)&listen_addr,
+                                            sizeof(listen_addr));
+    if (status == UCS_ERR_INVALID_ADDR) {
+        UCS_TEST_SKIP_R("cannot listen to " + ucs::sockaddr_to_str(&listen_addr));
+    }
+
+    ucp_ep_params_t ep_params = ucp_test::get_ep_params();
+    ep_params.field_mask      |= UCP_EP_PARAM_FIELD_FLAGS |
+                                 UCP_EP_PARAM_FIELD_SOCK_ADDR |
+                                 UCP_EP_PARAM_FIELD_ERR_HANDLING_MODE;
+    ep_params.err_mode         = UCP_ERR_HANDLING_MODE_PEER;
+    ep_params.flags            = UCP_EP_PARAMS_FLAGS_CLIENT_SERVER;
+    ep_params.sockaddr.addr    = (struct sockaddr*)&listen_addr;
+    ep_params.sockaddr.addrlen = sizeof(listen_addr);
+    sender().connect(&receiver(), ep_params);
+
+    tag_send(sender(), receiver());
+
+    /* wait for reverse ep to appear */
```

Comment:
@yosefe this is waiting for the ep on the listener side to be created? when the server's cb is invoked?

Reply:
yes